src/Polynomials.jl

```diff
@@ -363,6 +365,7 @@ p(4)   # 57 = 1 + 2*4 + 3*4^2
 """
 function polyval{T,S}(p::Poly{T}, x::S)
     R = promote_type(T,S)
+    T<:AbstractFloat && any(isnan(p.a)) && convert(R, NaN)
```

Comment:
The thing is there are a couple of tricks done in `Base`, as far as I can see, to avoid throwing errors. I was also confused when I read the documentation on `isnan`, but then, `isnan(x::Real) = false` makes sense. Since `NaN` is defined in floating point arithmetic, we should not be allowed to obtain `NaN` with integral types, as there are no `Inf`s defined for integral types, but `typemin` and `typemax`, instead.

The same trick applies for `x = 1; y = [1, 2]; x == y; isequal(x, y)` in `Base`. Normally, we are comparing a scalar and a vector for equality, which is much like comparing apples and pears. Yet, it does not throw an error, but simply returns false.

Following the same analogy, I would opt for defining `isnan` for `Poly` objects and checking for `isnan` for all the coefficients.

We can also follow your approach (the second one). Define:

```julia
isnan{T<:AbstractFloat}(p::Poly{T}) = any(isnan(p.a))
isnan{T<:AbstractFloat}(p::Poly{Complex{T}}) = any(isnan(p.a))
isnan{T<:AbstractFloat}(p::Poly{Poly{T}}) = any(isnan(p.a))
isnan{T<:AbstractFloat}(p::Poly{Poly{Complex{T}}}) = any(isnan(p.a))
isnan{T}(::Poly{T}) = false
```

Then we can check for `isnan(p)`, instead, in `polyval`, `polyder` and `polyint`. However, this is very dirty and meaningless, in my opinion, compared to assuming/requiring that custom `T<:Number` types have their `isnan`s implemented.

I don't know... Just thinking out loud...